File: android-native/KyouNoOgatore/app/src/main/java/jp/ogatore/kyouno/KyonoTypeArt.kt
package jp.ogatore.kyouno

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Fill
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.dp

// フォント適用漏れ・キャラ/タイプ画像の欠落修正タスク(TASK-C2-2026-07-26-visual-parity-fonts-characters.md)
// §3 かたさタイプの画像: app-quiz.js:3-7 TYPE_ART(koka/ashi/robotのインラインSVG)・
// index.html:1437 TYPE_IMG(momo/kenko/yawaraのPNG)の1:1移植。診断結果画面のタイプ名上部
// (Web版 .type-illust、104x104)に表示する。既存のKyonoIcons.ktと同じ要領でSVGはCanvas直書き。
private val TYPE_IMG_RES = mapOf(
    "momo" to "type_momo",
    "kenko" to "type_kenko",
    "yawara" to "type_yawara",
)

@Composable
fun KyonoTypeArt(typeKey: String, modifier: Modifier = Modifier) {
    val resName = TYPE_IMG_RES[typeKey]
    if (resName != null) {
        val context = LocalContext.current
        val resId = remember(resName) { context.resources.getIdentifier(resName, "drawable", context.packageName) }
        if (resId != 0) {
            Image(painter = painterResource(id = resId), contentDescription = typeKey, modifier = modifier.size(104.dp))
        }
    } else {
        Canvas(modifier.size(104.dp)) {
            val s = size.width / 96f
            fun pt(x: Float, y: Float) = Offset(x * s, y * s)
            val ink = Color(0xFF3A3A35)
            when (typeKey) {
                "koka" -> {
                    // app-quiz.js TYPE_ART.koka(開かずのトビラ=扉)の1:1移植。
                    val door = Path().apply {
                        addRoundRect(androidx.compose.ui.geometry.RoundRect(pt(24f, 12f).x, pt(24f, 12f).y, pt(72f, 86f).x, pt(72f, 86f).y, CornerRadius(6f * s)))
                    }
                    drawPath(door, Color(0xFFD9A066), style = Fill)
                    drawPath(door, ink, style = Stroke(3f * s))
                    val panelTop = Path().apply {
                        addRoundRect(androidx.compose.ui.geometry.RoundRect(pt(31f, 20f).x, pt(31f, 20f).y, pt(65f, 46f).x, pt(65f, 46f).y, CornerRadius(4f * s)))
                    }
                    val panelBottom = Path().apply {
                        addRoundRect(androidx.compose.ui.geometry.RoundRect(pt(31f, 52f).x, pt(31f, 52f).y, pt(65f, 78f).x, pt(65f, 78f).y, CornerRadius(4f * s)))
                    }
                    val panelFill = Color(0xFFE8B87E)
                    val panelBorder = Color(0xFFB4805A)
                    drawPath(panelTop, panelFill, style = Fill); drawPath(panelTop, panelBorder, style = Stroke(2f * s))
                    drawPath(panelBottom, panelFill, style = Fill); drawPath(panelBottom, panelBorder, style = Stroke(2f * s))
                    drawCircle(ink, radius = 3.5f * s, center = pt(63f, 49f))
                    drawCircle(ink, radius = 1.8f * s, center = pt(39f, 33f))
                    drawCircle(ink, radius = 1.8f * s, center = pt(53f, 33f))
                    val smile = Path().apply {
                        moveTo(pt(42f, 39f).x, pt(42f, 39f).y)
                        quadraticBezierTo(pt(46f, 41.5f).x, pt(46f, 41.5f).y, pt(50f, 39f).x, pt(50f, 39f).y)
                    }
                    drawPath(smile, ink, style = Stroke(2.2f * s, cap = StrokeCap.Round))
                }
                "ashi" -> {
                    // app-quiz.js TYPE_ART.ashi(棒立ちペンギン)の1:1移植。
                    drawOval(Color(0xFF4E5A6E), topLeft = Offset(pt(24f, 24f).x, pt(24f, 24f).y), size = androidx.compose.ui.geometry.Size(48f * s, 60f * s))
                    drawOval(ink, topLeft = Offset(pt(24f, 24f).x, pt(24f, 24f).y), size = androidx.compose.ui.geometry.Size(48f * s, 60f * s), style = Stroke(3f * s))
                    drawOval(Color.White, topLeft = Offset(pt(33f, 40f).x, pt(33f, 40f).y), size = androidx.compose.ui.geometry.Size(30f * s, 40f * s))
                    drawCircle(ink, radius = 2.2f * s, center = pt(41f, 40f))
                    drawCircle(ink, radius = 2.2f * s, center = pt(55f, 40f))
                    val beak = Path().apply {
                        moveTo(pt(44f, 47f).x, pt(44f, 47f).y); lineTo(pt(48f, 50f).x, pt(48f, 50f).y); lineTo(pt(52f, 47f).x, pt(52f, 47f).y); close()
                    }
                    val beakColor = Color(0xFFF5A25D)
                    drawPath(beak, beakColor, style = Fill)
                    drawPath(beak, ink, style = Stroke(2f * s, join = StrokeJoin.Round))
                    val footL = Path().apply {
                        moveTo(pt(34f, 82f).x, pt(34f, 82f).y)
                        quadraticBezierTo(pt(38f, 85f).x, pt(38f, 85f).y, pt(43f, 83f).x, pt(43f, 83f).y)
                    }
                    val footR = Path().apply {
                        moveTo(pt(62f, 82f).x, pt(62f, 82f).y)
                        quadraticBezierTo(pt(58f, 85f).x, pt(58f, 85f).y, pt(53f, 83f).x, pt(53f, 83f).y)
                    }
                    drawPath(footL, beakColor, style = Stroke(4f * s, cap = StrokeCap.Round))
                    drawPath(footR, beakColor, style = Stroke(4f * s, cap = StrokeCap.Round))
                    val wingL = Path().apply {
                        moveTo(pt(24f, 44f).x, pt(24f, 44f).y)
                        quadraticBezierTo(pt(20f, 52f).x, pt(20f, 52f).y, pt(26f, 58f).x, pt(26f, 58f).y)
                    }
                    val wingR = Path().apply {
                        moveTo(pt(72f, 44f).x, pt(72f, 44f).y)
                        quadraticBezierTo(pt(76f, 52f).x, pt(76f, 52f).y, pt(70f, 58f).x, pt(70f, 58f).y)
                    }
                    drawPath(wingL, ink, style = Stroke(3f * s, cap = StrokeCap.Round))
                    drawPath(wingR, ink, style = Stroke(3f * s, cap = StrokeCap.Round))
                }
                "robot" -> {
                    // app-quiz.js TYPE_ART.robot(ガチガチロボット)の1:1移植。
                    val head = Path().apply {
                        addRoundRect(androidx.compose.ui.geometry.RoundRect(pt(24f, 26f).x, pt(24f, 26f).y, pt(72f, 68f).x, pt(72f, 68f).y, CornerRadius(10f * s)))
                    }
                    val headFill = Color(0xFFC7D3DE)
                    drawPath(head, headFill, style = Fill)
                    drawPath(head, ink, style = Stroke(3f * s))
                    drawLine(ink, pt(48f, 26f), pt(48f, 14f), strokeWidth = 3f * s, cap = StrokeCap.Round)
                    drawCircle(Color(0xFFFF8A70), radius = 4f * s, center = pt(48f, 12f))
                    drawCircle(ink, radius = 4f * s, center = pt(48f, 12f), style = Stroke(2.5f * s))
                    val eyeL = Path().apply { addRoundRect(androidx.compose.ui.geometry.RoundRect(pt(33f, 38f).x, pt(33f, 38f).y, pt(43f, 48f).x, pt(43f, 48f).y, CornerRadius(3f * s))) }
                    val eyeR = Path().apply { addRoundRect(androidx.compose.ui.geometry.RoundRect(pt(53f, 38f).x, pt(53f, 38f).y, pt(63f, 48f).x, pt(63f, 48f).y, CornerRadius(3f * s))) }
                    drawPath(eyeL, Color.White, style = Fill); drawPath(eyeL, ink, style = Stroke(2.5f * s))
                    drawPath(eyeR, Color.White, style = Fill); drawPath(eyeR, ink, style = Stroke(2.5f * s))
                    drawCircle(ink, radius = 1.8f * s, center = pt(38f, 43f))
                    drawCircle(ink, radius = 1.8f * s, center = pt(58f, 43f))
                    drawLine(ink, pt(40f, 57f), pt(56f, 57f), strokeWidth = 2.5f * s, cap = StrokeCap.Round)
                    val base = Path().apply { addRoundRect(androidx.compose.ui.geometry.RoundRect(pt(30f, 72f).x, pt(30f, 72f).y, pt(66f, 82f).x, pt(66f, 82f).y, CornerRadius(5f * s))) }
                    drawPath(base, Color(0xFFE4EAF0), style = Fill)
                    drawPath(base, ink, style = Stroke(2.5f * s))
                }
            }
        }
    }
}

File: ios-native/KyouNoOgatore/KyouNoOgatore/KyonoTypeArt.swift
//
//  KyonoTypeArt.swift
//  KyouNoOgatore
//
//  フォント適用漏れ・キャラ/タイプ画像の欠落修正タスク(TASK-C2-2026-07-26-visual-parity-fonts-characters.md)
//  §3 かたさタイプの画像: app-quiz.js:3-7 TYPE_ART(koka/ashi/robotのインラインSVG)・
//  index.html:1437 TYPE_IMG(momo/kenko/yawaraのPNG)の1:1移植。診断結果画面のタイプ名上部
//  (Web版 .type-illust、104x104)に表示する。既存のKyonoIcons.swiftと同じ要領でSVGはCanvas直書き。
//
//  TestFlight実機フィードバックE1(2026-07-29): 6体ぶんのPNGが揃ったため、koka/ashi/robotも
//  PNG優先の対象に含めた。Canvas描画(下のGroup else節)はコードとしては残し、辞書のキーに無い
//  タイプが来たとき、またはPNG読み込み自体が失敗したとき(index.html:2624 TYPE_IMG[key]?PNG:SVG
//  と同じPNG優先・SVGフォールバックの考え方)の保険として使う。このプロジェクトはウィジェット
//  (B6)・検索サムネイル(D1)と、バンドル画像が実機で静かに読み込めなくなる欠陥を2度経験して
//  いるため、Canvas描画を消して二重管理を無くすより、フォールバックとして残す判断にした
//  (alan5の「消す」提案とは異なる判断・理由はここに明記)。

import SwiftUI

// TestFlight実機フィードバックE1(2026-07-29): alan5の当初の見立て(「koka/ashi/robotは
// 表示側が無い」)は誤りで、実際にはこの下のCanvas描画として既に実装済みだった。本当の作業は
// 「無いものを足す」ではなく「PNGが揃った(assets/type-{koka,ashi,robot}.png、承認済みコミット
// c32bf1f)ので、既存のPNG優先ロジックの対象に6体とも含める」こと。辞書に無いキーはCanvas描画
// にフォールバックする既存の分岐(下のGroup参照)を使うため、ここに追加するだけで済む。
private let typeImgNames: [String: String] = [
    "momo": "type-momo",
    "kenko": "type-kenko",
    "yawara": "type-yawara",
    "koka": "type-koka",
    "ashi": "type-ashi",
    "robot": "type-robot",
]

struct KyonoTypeArt: View {
    let typeKey: String

    var body: some View {
        Group {
            if let name = typeImgNames[typeKey], let uiImage = loadTypeArt(name) {
                Image(uiImage: uiImage).resizable().scaledToFit()
            } else {
                Canvas { ctx, size in
                    let s = size.width / 96
                    func pt(_ x: CGFloat, _ y: CGFloat) -> CGPoint { CGPoint(x: x * s, y: y * s) }
                    let ink = Color(hex: 0x3A3A35)
                    switch typeKey {
                    case "koka":
                        // app-quiz.js TYPE_ART.koka(開かずのトビラ=扉)の1:1移植。
                        let door = Path(roundedRect: CGRect(x: pt(24, 12).x, y: pt(24, 12).y, width: pt(48, 0).x, height: pt(0, 74).y), cornerRadius: 6 * s)
                        ctx.fill(door, with: .color(Color(hex: 0xD9A066)))
                        ctx.stroke(door, with: .color(ink), lineWidth: 3 * s)
                        let panelTop = Path(roundedRect: CGRect(x: pt(31, 20).x, y: pt(31, 20).y, width: pt(34, 0).x, height: pt(0, 26).y), cornerRadius: 4 * s)
                        let panelBottom = Path(roundedRect: CGRect(x: pt(31, 52).x, y: pt(31, 52).y, width: pt(34, 0).x, height: pt(0, 26).y), cornerRadius: 4 * s)
                        let panelFill = Color(hex: 0xE8B87E), panelBorder = Color(hex: 0xB4805A)
                        ctx.fill(panelTop, with: .color(panelFill)); ctx.stroke(panelTop, with: .color(panelBorder), lineWidth: 2 * s)
                        ctx.fill(panelBottom, with: .color(panelFill)); ctx.stroke(panelBottom, with: .color(panelBorder), lineWidth: 2 * s)
                        ctx.fill(Path(ellipseIn: CGRect(x: pt(63, 49).x - 3.5 * s, y: pt(63, 49).y - 3.5 * s, width: 7 * s, height: 7 * s)), with: .color(ink))
                        ctx.fill(Path(ellipseIn: CGRect(x: pt(39, 33).x - 1.8 * s, y: pt(39, 33).y - 1.8 * s, width: 3.6 * s, height: 3.6 * s)), with: .color(ink))
                        ctx.fill(Path(ellipseIn: CGRect(x: pt(53, 33).x - 1.8 * s, y: pt(53, 33).y - 1.8 * s, width: 3.6 * s, height: 3.6 * s)), with: .color(ink))
                        var smile = Path()
                        smile.move(to: pt(42, 39))
                        smile.addQuadCurve(to: pt(50, 39), control: pt(46, 41.5))
                        ctx.stroke(smile, with: .color(ink), style: StrokeStyle(lineWidth: 2.2 * s, lineCap: .round))
                    case "ashi":
                        // app-quiz.js TYPE_ART.ashi(棒立ちペンギン)の1:1移植。
                        let body = Path(ellipseIn: CGRect(x: pt(24, 24).x, y: pt(24, 24).y, width: 48 * s, height: 60 * s))
                        ctx.fill(body, with: .color(Color(hex: 0x4E5A6E)))
                        ctx.stroke(body, with: .color(ink), lineWidth: 3 * s)
                        let belly = Path(ellipseIn: CGRect(x: pt(33, 40).x, y: pt(33, 40).y, width: 30 * s, height: 40 * s))
                        ctx.fill(belly, with: .color(.white))
                        ctx.fill(Path(ellipseIn: CGRect(x: pt(41, 40).x - 2.2 * s, y: pt(41, 40).y - 2.2 * s, width: 4.4 * s, height: 4.4 * s)), with: .color(ink))
                        ctx.fill(Path(ellipseIn: CGRect(x: pt(55, 40).x - 2.2 * s, y: pt(55, 40).y - 2.2 * s, width: 4.4 * s, height: 4.4 * s)), with: .color(ink))
                        var beak = Path()
                        beak.move(to: pt(44, 47)); beak.addLine(to: pt(48, 50)); beak.addLine(to: pt(52, 47)); beak.closeSubpath()
                        let beakColor = Color(hex: 0xF5A25D)
                        ctx.fill(beak, with: .color(beakColor))
                        ctx.stroke(beak, with: .color(ink), style: StrokeStyle(lineWidth: 2 * s, lineJoin: .round))
                        var footL = Path(); footL.move(to: pt(34, 82)); footL.addQuadCurve(to: pt(43, 83), control: pt(38, 85))
                        var footR = Path(); footR.move(to: pt(62, 82)); footR.addQuadCurve(to: pt(53, 83), control: pt(58, 85))
                        ctx.stroke(footL, with: .color(beakColor), style: StrokeStyle(lineWidth: 4 * s, lineCap: .round))
                        ctx.stroke(footR, with: .color(beakColor), style: StrokeStyle(lineWidth: 4 * s, lineCap: .round))
                        var wingL = Path(); wingL.move(to: pt(24, 44)); wingL.addQuadCurve(to: pt(26, 58), control: pt(20, 52))
                        var wingR = Path(); wingR.move(to: pt(72, 44)); wingR.addQuadCurve(to: pt(70, 58), control: pt(76, 52))
                        ctx.stroke(wingL, with: .color(ink), style: StrokeStyle(lineWidth: 3 * s, lineCap: .round))
                        ctx.stroke(wingR, with: .color(ink), style: StrokeStyle(lineWidth: 3 * s, lineCap: .round))
                    case "robot":
                        // app-quiz.js TYPE_ART.robot(ガチガチロボット)の1:1移植。
                        let head = Path(roundedRect: CGRect(x: pt(24, 26).x, y: pt(24, 26).y, width: 48 * s, height: 42 * s), cornerRadius: 10 * s)
                        ctx.fill(head, with: .color(Color(hex: 0xC7D3DE)))
                        ctx.stroke(head, with: .color(ink), lineWidth: 3 * s)
                        var antenna = Path(); antenna.move(to: pt(48, 26)); antenna.addLine(to: pt(48, 14))
                        ctx.stroke(antenna, with: .color(ink), style: StrokeStyle(lineWidth: 3 * s, lineCap: .round))
                        let tip = Path(ellipseIn: CGRect(x: pt(48, 12).x - 4 * s, y: pt(48, 12).y - 4 * s, width: 8 * s, height: 8 * s))
                        ctx.fill(tip, with: .color(Color(hex: 0xFF8A70)))
                        ctx.stroke(tip, with: .color(ink), lineWidth: 2.5 * s)
                        let eyeL = Path(roundedRect: CGRect(x: pt(33, 38).x, y: pt(33, 38).y, width: 10 * s, height: 10 * s), cornerRadius: 3 * s)
                        let eyeR = Path(roundedRect: CGRect(x: pt(53, 38).x, y: pt(53, 38).y, width: 10 * s, height: 10 * s), cornerRadius: 3 * s)
                        ctx.fill(eyeL, with: .color(.white)); ctx.stroke(eyeL, with: .color(ink), lineWidth: 2.5 * s)
                        ctx.fill(eyeR, with: .color(.white)); ctx.stroke(eyeR, with: .color(ink), lineWidth: 2.5 * s)
                        ctx.fill(Path(ellipseIn: CGRect(x: pt(38, 43).x - 1.8 * s, y: pt(38, 43).y - 1.8 * s, width: 3.6 * s, height: 3.6 * s)), with: .color(ink))
                        ctx.fill(Path(ellipseIn: CGRect(x: pt(58, 43).x - 1.8 * s, y: pt(58, 43).y - 1.8 * s, width: 3.6 * s, height: 3.6 * s)), with: .color(ink))
                        var mouth = Path(); mouth.move(to: pt(40, 57)); mouth.addLine(to: pt(56, 57))
                        ctx.stroke(mouth, with: .color(ink), style: StrokeStyle(lineWidth: 2.5 * s, lineCap: .round))
                        let base = Path(roundedRect: CGRect(x: pt(30, 72).x, y: pt(30, 72).y, width: 36 * s, height: 10 * s), cornerRadius: 5 * s)
                        ctx.fill(base, with: .color(Color(hex: 0xE4EAF0)))
                        ctx.stroke(base, with: .color(ink), lineWidth: 2.5 * s)
                    default:
                        break
                    }
                }
            }
        }
        .frame(width: 104, height: 104)
    }
}

// PBXFileSystemSynchronizedRootGroup(Xcode26形式)はTypeArt/サブフォルダをビルド時にフラット化して
// バンドルルート直下へ配置するため、subdirectory指定なしで探す(DexView.swift loadCardArtと同じ考え方)。
private func loadTypeArt(_ name: String) -> UIImage? {
    guard let url = Bundle.main.url(forResource: name, withExtension: "png") else { return nil }
    return UIImage(contentsOfFile: url.path)
}
